### ISSUE TITLE
[reddit] Add support for reddit's own video host

### ISSUE THREAD

**Issue by goggle** [rg3/youtube-dl#13847]
### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2017.08.06*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2017.08.06**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones

### What is the purpose of your *issue*?
- [ ] Bug report (encountered problems with youtube-dl)
- [x] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add the `-v` flag to **your command line** you run youtube-dl with (`youtube-dl -v <your command line>`), copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v', 'https://www.reddit.com/r/videos/comments/6rrwyj/that_small_heart_attack/']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2017.08.06
[debug] Python version 3.6.2 - Linux-4.12.4-1-ARCH-x86_64-with-arch
[debug] exe versions: ffmpeg 3.3.3, ffprobe 3.3.3, rtmpdump 2.4
[debug] Proxy map: {}
[generic] that_small_heart_attack: Requesting header
WARNING: Falling back on generic information extractor.
[generic] that_small_heart_attack: Downloading webpage
[generic] that_small_heart_attack: Extracting information
ERROR: Unsupported URL: https://www.reddit.com/r/videos/comments/6rrwyj/that_small_heart_attack/
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/youtube_dl/YoutubeDL.py", line 776, in extract_info
    ie_result = ie.extract(url)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/common.py", line 433, in extract
    ie_result = self._real_extract(url)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/generic.py", line 2944, in _real_extract
    raise UnsupportedError(url)
youtube_dl.utils.UnsupportedError: Unsupported URL: https://www.reddit.com/r/videos/comments/6rrwyj/that_small_heart_attack/
```

---

### If the purpose of this *issue* is a *site support request* please provide all kinds of example URLs support for which should be included (replace following example URLs by **yours**):
- Single video: https://www.reddit.com/r/videos/comments/6rrwyj/that_small_heart_attack/ (or https://v.redd.it/zv89llsvexdz)
- Single video: https://www.reddit.com/r/holdmybeer/comments/6rn3e4/hmb_while_i_put_another_bottle_through_wood/
- Single video (without audio): https://www.reddit.com/r/NatureGifs/comments/6s0doa/bumblebee_lining_itself_up_to_land_on_a_swaying/

---

### Description of your *issue*, suggested solution and other information

Reddit has begun to host videos on its own on v.redd.it. It seems as they support to upload GIFs too and provide them as videos without audio through v.redd.it (if I understood this part correctly). It would be nice to have youtube-dl support for these videos.

**Comment by wongmjane** [comment]
As of now, it seems reddit's own video host uses (aka `v.redd.it`) uses DASH format, whose path to MPD manifest is:

`https://v.redd.it/[videoID]/DASHPlaylist.mpd` (for example [1])

It seems youtube-dl does recognize the formats available if I explicitly used the MPD link instead. For example:

```

# youtube-dl -v -F https://v.redd.it/31f0b5u64qez/DASHPlaylist.mpd

[debug] System config: []
[debug] User config: [u'--console-title', u'--hls-prefer-native', u'--audio-quality', u'0', u'--embed-subs', u'--embed-thumbnail', u'--add-metadata', u'--xattrs', u'--xattr-set-filesize', u'--write-sub', u'--prefer-ffmpeg', u'-f', u'bestvideo[ext=mp4]+bestaudio[ext=m4a]/bestvideo[ext=webm]+bestaudio[ext=webm]/best[ext=mp4]/best[ext=webm]/best']
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'-F', u'https://v.redd.it/31f0b5u64qez/DASHPlaylist.mpd']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2017.08.06
[debug] Python version 2.7.10 - Darwin-17.0.0-x86_64-i386-64bit
[debug] exe versions: ffmpeg 3.3.3, ffprobe 3.3.3, rtmpdump 2.4
[debug] Proxy map: {}
[generic] DASHPlaylist: Requesting header
WARNING: Falling back on generic information extractor.
[generic] DASHPlaylist: Downloading webpage
[generic] DASHPlaylist: Extracting information
[info] Available formats for DASHPlaylist:
format code           extension  resolution note
DASHPlaylist-AUDIO-1  m4a        audio only DASH audio  130k , mp4a.40.2 (48000Hz)
DASHPlaylist-VIDEO-2  mp4        240x240    DASH video  559k , avc1.4d401e, 30fps, video only
DASHPlaylist-VIDEO-1  mp4        360x360    DASH video 1104k , avc1.4d401e, 30fps, video only (best)

```

As shown above, it falls back to the generic extractor due to lack of current support to such host. And it shows the DASH video and DASH audio as if they are separate files.

[1] https://v.redd.it/on3hgdzlo1ez/DASHPlaylist.mpd

**Comment by wongmjane** [comment]
TL;DR / Workaround:

Append `/DASHPlaylist.mpd` at the end of that `v.redd.it` permalink, as the following format:

`https://v.reddit.com/[id]/DASHPlaylist.mpd`

For example, https://v.redd.it/on3hgdzlo1ez/DASHPlaylist.mpd works for youtube-dl but https://v.redd.it/on3hgdzlo1ez doesn't.

**Comment by goggle** [comment]
@wongmjane Thanks for your reply!
I've figured this out too. 
It is easy to write an information extractor for URLs of the form `https://v.reddit.com/[id]`.

The question is, should we support URLs of the form 
```
https://www.reddit.com/r/[subreddit]/comments/[post_id]/[post_name]/
```
?
I think we should, because that's the only reasonable way of using youtube-dl to download content from v.redd.it.

The only problem I see, is that we cannot really decide from URL of such a post, if it is a video post or not, e.g. https://www.reddit.com/r/videos/comments/6slvkg/how_to_destroy_a_man/ is such a video post, but https://www.reddit.com/r/videos/comments/6rfcyt/were_taking_part_in_the_video_beta/ is not. 

My question to the maintainers is, what is the correct youtube-dl way to do this? I've seen other information extractors, which just throw an error and tell the user to fill in a bug report, although it is not a bug, since the requested site does not contain a video. Isn't there a better way to handle such a situation?
